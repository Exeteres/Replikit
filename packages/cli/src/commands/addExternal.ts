import { program } from "commander";
import {
    setupConfiguration,
    getProjectManager,
    createExternalProjectName,
    logger,
    validateDirectoryName
} from "@replikit/cli";
import { prompt } from "inquirer";
import { pathExists, ensureDir } from "fs-extra";
import { join } from "path";

interface FirstPromptResult {
    url: string;
    name: string;
}

interface SecondPromptResult {
    modules: string[];
}

interface AdditionalPromptResult {
    postfix: string;
}

const command = program
    .command("add-external")
    .description("Add an external repository with Replikit modules");
command.action(async options => {
    await setupConfiguration(options.config);
    const project = await getProjectManager();
    const firstResult = await prompt<FirstPromptResult>([
        {
            type: "input",
            name: "url",
            message: "Repository url"
        },
        {
            type: "input",
            name: "name",
            message: "Name",
            default: (result: FirstPromptResult) => createExternalProjectName(result.url),
            validate: validateDirectoryName
        }
    ]);
    try {
        await ensureDir(project.externalPath);
        await project.git.addSubmodule(firstResult.url, firstResult.name);
    } catch (e) {
        logger.fatal("Error while adding the external repository", e);
    }
    let path = firstResult.name;
    const modulesPath = join(project.externalPath, firstResult.name, "modules");
    const modulesPathExists = await pathExists(modulesPath);
    if (!modulesPathExists) {
        const additionalResult = await prompt<AdditionalPromptResult>({
            type: "input",
            name: "postfix",
            message: "Relative path to Replikit root directory in the repository",
            validate: async (input: string) => {
                const modulesPath = join(project.externalPath, firstResult.name, input, "modules");
                const modulesPathExists = await pathExists(modulesPath);
                return modulesPathExists || 'Specified directory doesn\'t contain "modules"';
            }
        });
        path = join(firstResult.name, additionalResult.postfix);
    }
    try {
        await project.addExternalRepo(path);
        const allModules = await project.getExternalModuleNames(path);
        logger.info("Repository submodule added");
        const secondResult = await prompt<SecondPromptResult>([
            {
                type: "checkbox",
                name: "modules",
                message: "External modules",
                choices: allModules,
                validate: (input: string[]) => !!input.length || "No modules selected"
            }
        ]);
        await project.addLocalModules(secondResult.modules);
        await project.resolveExternalDependencies(path);
        logger.info("External modules added successfuly");
    } catch (e) {
        logger.fatal("Error while adding the external repository", e);
    }
});
