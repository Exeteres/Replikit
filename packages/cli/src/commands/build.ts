import { program } from "commander";
import {
    loadConfiguration,
    createWebpackConfiguration,
    getProjectManager,
    logger
} from "@replikit/cli";
import { updateConfig, config } from "@replikit/core";
import webpack from "webpack";
import { inspect } from "util";
import { remove, pathExists } from "fs-extra";
import { CliConfiguration } from "@replikit/cli/typings";

const command = program
    .command("build")
    .option("--print-config")
    .option("--development")
    .description("Build all modules to production-ready bundle");
command.action(async options => {
    const newConfig = await loadConfiguration(options.config);
    updateConfig(newConfig);

    const projectManager = await getProjectManager();
    const webpackConfig = createWebpackConfiguration(
        projectManager,
        config.cli as CliConfiguration,
        options.development
    );

    if (options.printConfig) {
        const message = inspect(webpackConfig, {
            showHidden: false,
            depth: null,
            colors: true
        });
        return console.log(message);
    }

    if (await pathExists(webpackConfig.output!.path!)) {
        await remove(webpackConfig.output!.path!);
    }

    webpack(webpackConfig, (error, stats) => {
        if (error) {
            return logger.error("Error while webpack compilation", error);
        }

        const message = stats.toString({
            errors: true,
            colors: true,
            all: false,
            assets: true,
            hash: true,
            version: true,
            builtAt: true
        });
        console.log(message.replace(/\[32m/gm, "[34m"));
    });
});
