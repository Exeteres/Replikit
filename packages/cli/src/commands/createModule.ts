import {
    getProjectManager,
    loadConfiguration,
    validateDirectoryName,
    availableModules,
    createModule,
    logger
} from "@replikit/cli";
import { program } from "commander";
import { prompt } from "inquirer";
import { updateConfig } from "@replikit/core";

const command = program
    .command("create-module")
    .description("Create a new module");
command.action(async options => {
    const config = await loadConfiguration(options.config);
    updateConfig(config);
    const result = await prompt([
        {
            type: "input",
            name: "moduleName",
            message: "Module name",
            validate: validateDirectoryName
        },
        {
            type: "confirm",
            name: "addLogger",
            message: "Do you want to add a logger to the module?"
        },
        {
            type: "checkbox",
            name: "modules",
            message: "Modules",
            choices: availableModules
        }
    ]);
    const manager = await getProjectManager();
    try {
        await createModule(
            manager,
            result.moduleName,
            result.modules,
            result.addLogger
        );
        logger.info("Module created successfuly");
    } catch (e) {
        logger.fatal("Error while creating the module", e);
    }
});
