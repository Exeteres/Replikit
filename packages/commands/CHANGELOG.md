# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.6.0](https://github.com/Exeteres/Replikit/compare/v0.5.3...v0.6.0) (2020-06-21)

**Note:** Version bump only for package @replikit/commands





## [0.5.3](https://github.com/Exeteres/Replikit/compare/v0.5.2...v0.5.3) (2020-06-20)

**Note:** Version bump only for package @replikit/commands





## [0.5.2](https://github.com/Exeteres/Replikit/compare/v0.5.1...v0.5.2) (2020-06-20)

**Note:** Version bump only for package @replikit/commands





# 0.5.0 (2020-06-20)


### Bug Fixes

* **commands:** Fix validation of text parameter ([4ab05f8](https://github.com/Exeteres/Replikit/commit/4ab05f84120449aeb0a8b9d2e38d9e9c7a23bc0e))


### Features

* **commands:** Rename multiline parameter to text parameter and make it possible to work with a rest parameter or instead of it ([6b2338d](https://github.com/Exeteres/Replikit/commit/6b2338d88b2b5452eacf39b690b797d30a6bf81f))
* **help:** Add help package ([68b50e9](https://github.com/Exeteres/Replikit/commit/68b50e9d04561ad037bcd2abcb5fbeffd4f2c664))






# [0.4.0](https://github.com/Exeteres/Replikit/compare/v0.3.0...v0.4.0) (2020-06-19)


### Bug Fixes

* **commands:** Fix validation of text parameter ([fa892f2](https://github.com/Exeteres/Replikit/commit/fa892f24eadf262401da9fd7f5a8dcb2750b0fdf))





# [0.3.0](https://github.com/Exeteres/Replikit/compare/v0.2.0...v0.3.0) (2020-06-14)


### Features

* **commands:** Rename multiline parameter to text parameter and make it possible to work with a rest parameter or instead of it ([44c5b69](https://github.com/Exeteres/Replikit/commit/44c5b69cfca82a0a7280593e6bef8b2028ca2ae0))





# [0.2.0](https://github.com/Exeteres/Replikit/compare/v0.1.0...v0.2.0) (2020-06-13)

**Note:** Version bump only for package @replikit/commands
