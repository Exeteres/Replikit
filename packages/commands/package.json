{
    "name": "@replikit/commands",
    "version": "0.18.0",
    "license": "MIT",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "typings"
    ],
    "scripts": {
        "build": "tsc -p tsconfig.build.json"
    },
    "dependencies": {
        "@replikit/core": "^0.18.0",
        "@replikit/i18n": "^0.18.0",
        "@replikit/router": "^0.18.0"
    }
}
