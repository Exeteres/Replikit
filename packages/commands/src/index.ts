/// <reference path="../typings/index.d.ts" />

export * from "./commandsLocale";
export * from "./usageRenderer";
export * from "./middlewareStage";
export * from "./middlewareRouter";
export * from "./errors";
export * from "./converterStorage";
export * from "./converterBuilder";
export * from "./commandStorage";
export * from "./commandBuilder";
export * from "./converters";

import "./locales/en";
import "./locales/ru";
import "./startup";
