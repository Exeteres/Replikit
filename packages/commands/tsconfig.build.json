{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "dist",
        "paths": {
            "@replikit/commands": ["commands/src"]
        }
    },
    "include": ["src"]
}
