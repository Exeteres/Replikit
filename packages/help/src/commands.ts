import { command, commands } from "@replikit/commands";
import { descriptions } from "@replikit/help";
import { fromCode } from "@replikit/messages";

command("help")
    .handler(context => {
        const message = descriptions.render(
            commands.getCommands(),
            context.account.language
        );
        return fromCode(message);
    })
    .register();
