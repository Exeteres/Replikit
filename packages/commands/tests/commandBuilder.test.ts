import { MiddlewareStage } from "@replikit/commands";
import { createTestManager } from "@replikit/test-utils";

describe("CommandBuilder", () => {
    it("should build a complex command", () => {
        const { command } = createTestManager();
        const result = command("test", "t")
            .required("p1", String)
            .optional("p2", Number)
            .optional("p3", Boolean, { default: true })
            .rest("p4", Number)
            .multiline("p5", true, true)
            .use(MiddlewareStage.AfterResolution, context => void context)
            .use({
                stage: MiddlewareStage.BeforeResolution,
                handler: context => void context
            })
            .commands(command("subcommand"))
            .build();
        expect(result).toMatchSnapshot();
    });

    it("should register a constructed command in a CommandStorage", () => {
        const { testManager, command } = createTestManager();
        command("test").register();
        expect(testManager.commands).toMatchSnapshot();
    });
});
