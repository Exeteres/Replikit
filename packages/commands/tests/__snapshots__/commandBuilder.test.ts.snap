// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandBuilder should build a complex command 1`] = `
Object {
  "aliases": Array [
    "t",
  ],
  "commands": Array [
    Object {
      "aliases": Array [],
      "commands": Array [],
      "name": "subcommand",
      "params": Array [],
      "parent": [Circular],
      "requiredCount": 0,
    },
  ],
  "middlewareRouter": MiddlewareRouter {
    "chainMap": Map {
      2 => HandlerChain {
        "firstHandler": HandlerWrapper {
          "handler": [Function],
        },
        "lastHandler": HandlerWrapper {
          "handler": [Function],
        },
      },
      1 => HandlerChain {
        "firstHandler": HandlerWrapper {
          "handler": [Function],
        },
        "lastHandler": HandlerWrapper {
          "handler": [Function],
        },
      },
    },
  },
  "name": "test",
  "params": Array [
    Object {
      "isString": true,
      "name": "p1",
      "options": Object {},
    },
    Object {
      "converter": Object {
        "validator": [Function],
      },
      "isString": false,
      "name": "p2",
      "optional": true,
      "options": Object {},
      "type": [Function],
    },
    Object {
      "converter": Object {
        "validator": [Function],
      },
      "isString": false,
      "name": "p3",
      "optional": true,
      "options": Object {
        "default": true,
      },
      "type": [Function],
    },
  ],
  "requiredCount": 1,
  "rest": Object {
    "converter": Object {
      "validator": [Function],
    },
    "isString": false,
    "name": "p4",
    "options": Object {},
    "type": [Function],
  },
  "text": Object {
    "name": "p5",
    "skipValidation": true,
    "splitLines": true,
  },
}
`;

exports[`CommandBuilder should register a constructed command in a CommandStorage 1`] = `
CommandStorage {
  "commands": Array [
    Object {
      "aliases": Array [],
      "commands": Array [],
      "name": "test",
      "params": Array [],
      "requiredCount": 0,
      "usage": "/test {}",
    },
  ],
  "prefix": "/",
}
`;
